### ISSUE TITLE
panic: runtime error: invalid memory address or nil pointer dereference [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xbbe71f]

### ISSUE THREAD

**Issue by illagrenan** [semi-technologies/weaviate#1900]
👉 The environment is the same as in this issue: https://github.com/semi-technologies/weaviate/issues/1897 Versions tested: `v1.11.0` and `v1.12.1`.

After several successful GET queries, the container crashes with this error:

```
***","RawPath":"","ForceQuery":false,"RawQuery":"","Fragment":"","RawFragment":""}}
{"action":"restapi_request","level":"debug","method":"GET","msg":"received HTTP request","time":"2022-04-07T10:56:41Z","url":{"Scheme":"","Opaque":"","User":null,"Host":"","Path":"/v1/schema/***","RawPath":"","ForceQuery":false,"RawQuery":"","Fragment":"","RawFragment":""}}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (628002) without a tombstone, tombstone was added","node_id":628002,"time":"2022-04-07T10:57:49Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (628550) without a tombstone, tombstone was added","node_id":628550,"time":"2022-04-07T10:57:51Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (628646) without a tombstone, tombstone was added","node_id":628646,"time":"2022-04-07T10:57:51Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (629032) without a tombstone, tombstone was added","node_id":629032,"time":"2022-04-07T10:57:54Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (629068) without a tombstone, tombstone was added","node_id":629068,"time":"2022-04-07T10:57:54Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (629097) without a tombstone, tombstone was added","node_id":629097,"time":"2022-04-07T10:57:54Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (629279) without a tombstone, tombstone was added","node_id":629279,"time":"2022-04-07T10:57:55Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (629370) without a tombstone, tombstone was added","node_id":629370,"time":"2022-04-07T10:57:55Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (629507) without a tombstone, tombstone was added","node_id":629507,"time":"2022-04-07T10:57:55Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (629673) without a tombstone, tombstone was added","node_id":629673,"time":"2022-04-07T10:57:56Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (630165) without a tombstone, tombstone was added","node_id":630165,"time":"2022-04-07T10:57:59Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (630334) without a tombstone, tombstone was added","node_id":630334,"time":"2022-04-07T10:58:00Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (630401) without a tombstone, tombstone was added","node_id":630401,"time":"2022-04-07T10:58:00Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (630453) without a tombstone, tombstone was added","node_id":630453,"time":"2022-04-07T10:58:00Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (630470) without a tombstone, tombstone was added","node_id":630470,"time":"2022-04-07T10:58:00Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (630515) without a tombstone, tombstone was added","node_id":630515,"time":"2022-04-07T10:58:01Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (630516) without a tombstone, tombstone was added","node_id":630516,"time":"2022-04-07T10:58:01Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (630909) without a tombstone, tombstone was added","node_id":630909,"time":"2022-04-07T10:58:02Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (630910) without a tombstone, tombstone was added","node_id":630910,"time":"2022-04-07T10:58:02Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (630960) without a tombstone, tombstone was added","node_id":630960,"time":"2022-04-07T10:58:02Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (631244) without a tombstone, tombstone was added","node_id":631244,"time":"2022-04-07T10:58:03Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (631257) without a tombstone, tombstone was added","node_id":631257,"time":"2022-04-07T10:58:03Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (631731) without a tombstone, tombstone was added","node_id":631731,"time":"2022-04-07T10:58:07Z"}
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (631890) without a tombstone, tombstone was added","node_id":631890,"time":"2022-04-07T10:58:08Z"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xbbe71f]

goroutine 1111 [running]:
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*vertex).isUnderMaintenance(0x0, 0xf0100)
    /go/src/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/index.go:362 +0x3f
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).findBestEntrypointForNode(0xc0002a7b80, 0x3, 0x0, 0x1c2c4, 0xc1c03a3000, 0x600, 0x600, 0x0, 0xc810e8ef30, 0x0)
    /go/src/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/index.go:327 +0x234
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).reassignNeighborsOf(0xc0002a7b80, 0xc810e8ef30, 0xc060422000, 0x0)
    /go/src/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/delete.go:206 +0x365
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).CleanUpTombstonedNodes(0xc0002a7b80, 0xc0f4a16620, 0x0)
    /go/src/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/delete.go:124 +0x75
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).registerTombstoneCleanup.func1(0xc0002a7b80)
    /go/src/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/startup.go:134 +0xda
created by github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).registerTombstoneCleanup
    /go/src/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/startup.go:127 +0x53
Stream closed EOF for weaviate/weaviate-statefulset-0 (weaviate)
```

**Comment by etiennedi** [comment]
Thanks for the report. It seems like this is a follow-up to the other crash which already cleaned up some nodes, but not completely due to the previous panic. This is great for us to investigate. I can already see based on the stack trace above that there is a nil-check missing in one place. There could be other places. A possible way to narrow this down might be to push a container that fixes this check, see if it starts crashing in another place. If yes, we repeat until we have fixed all of them. Would that be OK for you?

**Comment by illagrenan** [comment]
Sounds like a plan. I agree.

**Comment by etiennedi** [comment]
Are you running on `arm64` or `amd64`? I'm planning to push the containers directly from my local machine since this is faster, but I just realized that if I push them from the Intel Mac they will be `amd64` if I push them from the M1 Mac they'll be `arm64`. Which one do you need?

**Comment by illagrenan** [comment]
I'm using `amd64` (Intel Xeon)

**Comment by etiennedi** [comment]
Awesome, in this case, this is the right image for you: `docker.io/semitechnologies/weaviate:1.12.1-gh-1900-iteration-2`. Let me know if the issue from above is indeed gone and which issue pops up now :-)

**Comment by illagrenan** [comment]
So far everything looks great. I see thousands of events like this:

```
...
{"action":"attach_tombstone_to_deleted_node","level":"info","msg":"found a deleted node (1089365) without a tombstone, tombstone was added","node_id":1089365,"time":"2022-04-07T13:14:51Z"}
...
```

After 20 minutes of uptime, Weaviate is still responsive and the API works as it should. 👍👍👍 gj!

**Comment by illagrenan** [comment]
Everything works for me. Thanks for solving the problem.

**Comment by etiennedi** [comment]
Released as part of v1.12.2.
